## Data Cleaning
---

- #### Step 1: import the necessary python packages
- #### Step 2: import the dataset
- #### Step 3: clean the data in the dataset
    - #### a. find all null values
    - #### b. impute or drop the missing data
    - #### c. check for duplicate values
- #### Step 4: store the cleaned data into a file with a readable format

In [3]:
import numpy as np
import pandas as pd

In [4]:
from kaggle.api.kaggle_api_extended import KaggleApi

# Initialize the Kaggle API
api = KaggleApi()

# Authenticate using the credentials in kaggle.json
try:
    api.authenticate()
    print("API authentication successful.")
except Exception as e:
    print(f"API authentication failed: {e}")

API authentication successful.


In [5]:
from kaggle.api.kaggle_api_extended import KaggleApi

# Initialize Kaggle API
api = KaggleApi()
api.authenticate()

# Define the dataset identifier and download path
dataset = 'brandao/Diabetes'  # dataset identifier
download_path = 'C:\\Users\\3sthe\\'

# Download the dataset
api.dataset_download_files(dataset, path=download_path, unzip=True)

Dataset URL: https://www.kaggle.com/datasets/brandao/Diabetes


In [6]:
data = pd.read_csv('diabetic_data.csv')
data
data.shape

(101766, 50)

In [7]:
# Replace "?" with NaN
data.replace("?", np.nan, inplace=True)

# print the total number of NaN values
print(data.isna().sum())

encounter_id                    0
patient_nbr                     0
race                         2273
gender                          0
age                             0
weight                      98569
admission_type_id               0
discharge_disposition_id        0
admission_source_id             0
time_in_hospital                0
payer_code                  40256
medical_specialty           49949
num_lab_procedures              0
num_procedures                  0
num_medications                 0
number_outpatient               0
number_emergency                0
number_inpatient                0
diag_1                         21
diag_2                        358
diag_3                       1423
number_diagnoses                0
max_glu_serum               96420
A1Cresult                   84748
metformin                       0
repaglinide                     0
nateglinide                     0
chlorpropamide                  0
glimepiride                     0
acetohexamide 

#### Insight: We can see from the above outputs that the following variables have missing values:
- #### race - 2273; weight - 98569; payer_code - 40256; medical_speciality - 49949; diag_1 - 21; diag_2 - 358; diag_3 - 1423; max_glu_serum - 96420; A1Cresult - 84748

In [9]:
# looking into 'race' column
print(data['race'].unique())

['Caucasian' 'AfricanAmerican' nan 'Other' 'Asian' 'Hispanic']


In [10]:
# Replace NaN in 'race' with Other
data['race'].fillna('Other', inplace=True)

In [11]:
print(data['race'].unique())

['Caucasian' 'AfricanAmerican' 'Other' 'Asian' 'Hispanic']


In [12]:
# looking into 'weight' column
print(data['weight'].unique())

[nan '[75-100)' '[50-75)' '[0-25)' '[100-125)' '[25-50)' '[125-150)'
 '[175-200)' '[150-175)' '>200']


In [13]:
# calculate the percentage of 'nan' values in the weight column
nan_weight_percentage = data['weight'].isna().mean() * 100
print('Current percentage of NaN values in weight:', round(nan_weight_percentage), '%')

Current percentage of NaN values in weight: 97 %


In [14]:
# replace 'nan' with 0

# convert string to integer in weight column using median values for each weight group
# create an age map respectively
weight_map = {'[75-100)': 88, '[50-75)': 64, '[0-25)': 14, '[100-125)': 114, '[25-50)': 39, '[125-150)': 139, '[175-200)': 189, '[150-175)': 164, '>200': 200}
# replace string in age column with median integer
data['weight'] = data['weight'].replace(weight_map)

In [15]:
# looking into 'weight' column
print(data['weight'].unique())

[ nan  88.  64.  14. 114.  39. 139. 189. 164. 200.]


In [16]:
# Impute missing values with the most frequent value
mode_weight = data['weight'].mode()[0]
data['weight'].fillna(mode_weight, inplace=True)

# check if there are anymore nan values
# calculate the percentage of 'nan' values in the medical_specialty column
nan_percentage_mode_weight = data['weight'].isna().mean() * 100
print('Updated percentage of NaN values in weight:', (nan_percentage_mode_weight), '%')

Updated percentage of NaN values in weight: 0.0 %


In [17]:
data.shape

(101766, 50)

In [18]:
# Impute missing values with the most frequent value
most_frequent_payer_code = data['medical_specialty'].mode()[0]
data['medical_specialty'].fillna(most_frequent_payer_code, inplace=True)

# check if there are anymore nan values
# calculate the percentage of 'nan' values in the medical_specialty column
nan_percentage_medSpecialty = data['medical_specialty'].isna().mean() * 100
print('Updated percentage of NaN values in medical_specialty:', (nan_percentage_medSpecialty), '%')

Updated percentage of NaN values in medical_specialty: 0.0 %


In [19]:
print(data['max_glu_serum'].unique())

[nan '>300' 'Norm' '>200']


In [20]:
data['max_glu_serum'] = data['max_glu_serum'].replace('>300', 300)
data['max_glu_serum'] = data['max_glu_serum'].replace('>200', 200)
data['max_glu_serum'] = data['max_glu_serum'].replace('Norm', 0)

In [21]:
print(data['max_glu_serum'].unique())

[ nan 300.   0. 200.]


In [22]:
print(data['max_glu_serum'].isna().sum())

96420


In [23]:
print(data['max_glu_serum'].value_counts())

max_glu_serum
0.0      2597
200.0    1485
300.0    1264
Name: count, dtype: int64


In [24]:
# Impute missing values with the most frequent value
most_frequent_max_glu_serum = data['max_glu_serum'].mode()[0]
data['max_glu_serum'].fillna(most_frequent_max_glu_serum, inplace=True)

# check if there are anymore nan values
# calculate the percentage of 'nan' values in the max_glu_serum column
nan_percentage_max_glu_serum = data['max_glu_serum'].isna().mean() * 100
print('Updated percentage of NaN values in max_glu_serum:', (nan_percentage_max_glu_serum), '%')

Updated percentage of NaN values in max_glu_serum: 0.0 %


In [25]:
print(data['max_glu_serum'].value_counts())

max_glu_serum
0.0      99017
200.0     1485
300.0     1264
Name: count, dtype: int64


In [26]:
# data.shape

In [27]:
# looking in to A1Cresult column
print(data['A1Cresult'].unique())

[nan '>7' '>8' 'Norm']


In [28]:
print(data['A1Cresult'].isna().sum())

84748


In [29]:
print(data['A1Cresult'].value_counts())

A1Cresult
>8      8216
Norm    4990
>7      3812
Name: count, dtype: int64


In [30]:
data['A1Cresult'] = data['A1Cresult'].replace('>8', 8)
data['A1Cresult'] = data['A1Cresult'].replace('>7', 7)
data['A1Cresult'] = data['A1Cresult'].replace('Norm', 0)

In [31]:
print(data['A1Cresult'].value_counts())

A1Cresult
8.0    8216
0.0    4990
7.0    3812
Name: count, dtype: int64


In [32]:
# Impute missing values with the most frequent value
most_frequent_A1Cresult = data['A1Cresult'].mode()[0]
data['A1Cresult'].fillna(most_frequent_A1Cresult, inplace=True)

# check if there are anymore nan values
# calculate the percentage of 'nan' values in the max_glu_serum column
nan_percentage_A1Cresult = data['A1Cresult'].isna().mean() * 100
print('Updated percentage of NaN values in A1Cresult:', (nan_percentage_A1Cresult), '%')

Updated percentage of NaN values in A1Cresult: 0.0 %


In [33]:
print(data['A1Cresult'].value_counts())

A1Cresult
8.0    92964
0.0     4990
7.0     3812
Name: count, dtype: int64


In [34]:
# print the total number of NaN values
# print(data.isna().sum())

In [35]:
# looking into payer code
print(data['payer_code'].value_counts())

payer_code
MC    32439
HM     6274
SP     5007
BC     4655
MD     3532
CP     2533
UN     2448
CM     1937
OG     1033
PO      592
DM      549
CH      146
WC      135
OT       95
MP       79
SI       55
FR        1
Name: count, dtype: int64


In [36]:
print(data['payer_code'].isna().sum())

40256


In [37]:
# fill nan values in payer code with mode value
# Calculate the most frequent value (mode) for imputation
mode_payer_code = data['payer_code'].mode()[0]

# Impute missing values with the mode value
data['payer_code'].fillna(mode_payer_code, inplace=True)

# Check if there are any remaining NaN values
nan_percentage_payer_code = data['payer_code'].isna().mean() * 100
print('Updated percentage of NaN values in payer_code:', nan_percentage_payer_code, '%')

Updated percentage of NaN values in payer_code: 0.0 %


In [38]:
# print(data['payer_code'].isna().sum())

In [39]:
print(data.isna().sum())

encounter_id                   0
patient_nbr                    0
race                           0
gender                         0
age                            0
weight                         0
admission_type_id              0
discharge_disposition_id       0
admission_source_id            0
time_in_hospital               0
payer_code                     0
medical_specialty              0
num_lab_procedures             0
num_procedures                 0
num_medications                0
number_outpatient              0
number_emergency               0
number_inpatient               0
diag_1                        21
diag_2                       358
diag_3                      1423
number_diagnoses               0
max_glu_serum                  0
A1Cresult                      0
metformin                      0
repaglinide                    0
nateglinide                    0
chlorpropamide                 0
glimepiride                    0
acetohexamide                  0
glipizide 

In [40]:
# looking into diagnoses columns
# fill nan values in payer code with mode value
# Calculate the most frequent value (mode) for imputation
mode_diag_1 = data['diag_1'].mode()[0]

# Impute missing values with the mode value
data['diag_1'].fillna(mode_diag_1, inplace=True)

# Check if there are any remaining NaN values
nan_percentage_diag_1 = data['diag_1'].isna().mean() * 100
print('Updated percentage of NaN values in diag_1:', nan_percentage_diag_1, '%')

Updated percentage of NaN values in diag_1: 0.0 %


In [41]:
# print(data['diag_1'].isna().sum())

In [42]:
# Calculate the most frequent value (mode) for imputation
mode_diag_2 = data['diag_2'].mode()[0]

# Impute missing values with the mode value
data['diag_2'].fillna(mode_diag_2, inplace=True)

# Check if there are any remaining NaN values
nan_percentage_diag_2 = data['diag_2'].isna().mean() * 100
print('Updated percentage of NaN values in diag_2:', nan_percentage_diag_2, '%')

Updated percentage of NaN values in diag_2: 0.0 %


In [43]:
# Calculate the most frequent value (mode) for imputation
mode_diag_3 = data['diag_3'].mode()[0]

# Impute missing values with the mode value
data['diag_3'].fillna(mode_diag_3, inplace=True)

# Check if there are any remaining NaN values
nan_percentage_diag_3 = data['diag_3'].isna().mean() * 100
print('Updated percentage of NaN values in diag_3:', nan_percentage_diag_3, '%')

Updated percentage of NaN values in diag_3: 0.0 %


In [44]:
print(data.isna().sum())

encounter_id                0
patient_nbr                 0
race                        0
gender                      0
age                         0
weight                      0
admission_type_id           0
discharge_disposition_id    0
admission_source_id         0
time_in_hospital            0
payer_code                  0
medical_specialty           0
num_lab_procedures          0
num_procedures              0
num_medications             0
number_outpatient           0
number_emergency            0
number_inpatient            0
diag_1                      0
diag_2                      0
diag_3                      0
number_diagnoses            0
max_glu_serum               0
A1Cresult                   0
metformin                   0
repaglinide                 0
nateglinide                 0
chlorpropamide              0
glimepiride                 0
acetohexamide               0
glipizide                   0
glyburide                   0
tolbutamide                 0
pioglitazo

In [90]:
# Check for duplicates in specific columns
duplicates_column = data.duplicated(subset=['encounter_id'])
print('Duplicate rows based on column1:\n', data[duplicates_column])

Duplicate rows based on column1:
 Empty DataFrame
Columns: [encounter_id, patient_nbr, race, gender, age, weight, admission_type_id, discharge_disposition_id, admission_source_id, time_in_hospital, payer_code, medical_specialty, num_lab_procedures, num_procedures, num_medications, number_outpatient, number_emergency, number_inpatient, diag_1, diag_2, diag_3, number_diagnoses, max_glu_serum, A1Cresult, metformin, repaglinide, nateglinide, chlorpropamide, glimepiride, acetohexamide, glipizide, glyburide, tolbutamide, pioglitazone, rosiglitazone, acarbose, miglitol, troglitazone, tolazamide, examide, citoglipton, insulin, glyburide-metformin, glipizide-metformin, glimepiride-pioglitazone, metformin-rosiglitazone, metformin-pioglitazone, change, diabetesMed, readmitted]
Index: []

[0 rows x 50 columns]


In [45]:
data.to_csv('diabetes_cleaned.csv', index=False)